src/tests/Microsoft.PowerFx.Interpreter.Tests/Scenarios/DataTableMarshallerProvider.cs

```diff
@@ -0,0 +1,123 @@
+using System.Collections.Generic;
+using System.Data;
+using Microsoft.PowerFx.Interpreter.Tests;
+using Microsoft.PowerFx.Types;
+
+namespace Microsoft.PowerFx.Tests
+{
+    /// <summary>
+    /// Wire up <see cref="DataTable"/> conversions to <see cref="TypeMarshallerCache.Marshal{T}(T)"/>.     
+    /// </summary>
+    internal class DataTableMarshallerProvider : IDynamicTypeMarshaller
+    {
+        public bool TryMarshal(TypeMarshallerCache cache, object value, out FormulaValue result)
+        {
+            if (value is DataTable dataTable)
+            {
+                result = new DataTableValue(dataTable);
+                return true;
+            }
+
+            result = null;
+            return false;
+        }
+    }
+
+    // Wrap a System.Data.DataTable as a Power Fx TableValue
+    // Marshal DataTable as a Table of Records. 
+    // Use column has type info if available; else marshal as untyped object.
+    // All marshalling is done lazily so we avoid copying the whole table.
+    internal class DataTableValue : CollectionTableValue<DataRow>
+    {
+        public DataTableValue(DataTable dataTable)
+            : base(ComputeType(dataTable), new DataTableWrapper(dataTable))
+        {
+        }
+
+        public static RecordType ComputeType(DataTable dataTable)
+        {
+            var recordType = new RecordType();
+            foreach (DataColumn column in dataTable.Columns)
+            {
+                if (!PrimitiveValueConversions.TryGetFormulaType(column.DataType, out var fxType))
+                {
+                    fxType = FormulaType.UntypedObject;
+                }
+
+                recordType = recordType.Add(column.ColumnName, fxType);
+            }
+
+            return recordType;
+        }
+
+        protected override DValue<RecordValue> Marshal(DataRow item)
+        {
+            var record = new DataRowRecordValue(RecordType, item);
+
+            return DValue<RecordValue>.Of(record);
+        }
+
+        // Wrap an individual DataRow of the DataTable as a Power Fx RecordValue
+        private class DataRowRecordValue : RecordValue
+        {
+            private readonly DataRow _row;
+
+            public DataRowRecordValue(RecordType type, DataRow row)
+                : base(type)
+            {
+                _row = row;
+            }
+
+            protected override bool TryGetField(FormulaType fieldType, string fieldName, out FormulaValue result)
+            {
+                var value = _row[fieldName];
+
+                // The expected fieldType was determined in ComputeType based on Column.DataType.
+                if (fieldType == FormulaType.UntypedObject)
+                {
+                    result = PrimitiveWrapperAsUnknownObject.New(value);
+                } 
+                else
+                {
+                    result = PrimitiveValueConversions.Marshal(value, fieldType);
+                }
+
+                return true;
```

Comment:
Always true?

Reply:
Good point. Line 77 would have failed. That's where the check should be.